proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,205 @@
+<!-- {% raw %} -->
+
+# Implementation of RootSignature in clang
+
+* Proposal: [0021](0021-root-signature-in-clang.md)
+* Author(s): [Xiang Li](https://github.com/python3kgae)
+* Sponsor: TBD
+* Status: **Under Consideration**
+
+
+## Introduction
+
+In HLSL, the [root signature](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signatures)
+defines what types of resources are bound to the graphics pipeline.
+
+A root signature can be specified in HLSL as a [string](https://learn.microsoft.com/en-us/windows/win32/direct3d12/specifying-root-signatures-in-hlsl#an-example-hlsl-root-signature).
+The string contains a collection of comma-separated clauses that describe root
+signature constituent components.
+
+## Motivation
+
+To support root signature in clang-dxc.
+
+## Proposed solution
+
+The root signature in a compiler begins with a string in the root signature
+attribute and ends with a serialized root signature in the DXContainer.
+
+To report errors as early as possible, root signature string parsing should
+occur in Sema.
+
+To ensure that a used resource has a binding in the root signature, this
+information should be accessible in the backend to make sure legal dxil is
+generated.
+
+A new AST node HLSLRootSignatureDecl will be added to represent the root
+signature in the AST.
+Assuming the cost to parse root signature string is cheap,
+HLSLRootSignatureDecl will follow common Clang approach which only save the
+string in the AST.
+
+A root signature will be parsed twice, once in Sema for diagnostic and once in
+clang code generation for generating the serialized root signature.
+The first parsing will check register overlap and run on all root signatures
+in the translation unit.
+The second parsing will calculate the offset for serialized root signature and
+only run on the root signature for the entry function or local/global root
+signature which used in SubobjectToExportsAssociation.
+
+HLSLRootSignatureDecl will have method to parse the string for diagnostic and
+generate the SerializedRS mentioned above.
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+
+For case compile to a standalone root signature blob, the
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into a global variable with struct type to express the layout and metadata to
+save things like static sampler, root flags, space and NumDescriptors in LLVM IR.
```

Comment:
Why a struct type? Why not metadata? Given the need to express more than just a packing of types it seems like metadata is better suited to this problem.

Reply:
The pro for use struct is that it saved the root signature in the serialized layout while you can still access the element when needed.
Then when emit serialized root signature, just pack the data into a vector of uint32_t.